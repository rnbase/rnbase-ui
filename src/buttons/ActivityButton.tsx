import React, { useEffect, useState } from 'react';
import { ActivityIndicator, Animated, Easing, StyleSheet, View } from 'react-native';

import Button, { ButtonProps } from './Button';

import { Stylized, WithThemeProps, withTheme } from '../theming';

interface ActivityButtonProps extends ButtonProps {
  busy?: boolean;
  indicatorColor?: string;
}

const toValue = (value?: boolean) => (value ? 1 : 0);

const ActivityButton: React.FC<Stylized<typeof createStyleSheet, ActivityButtonProps>> = ({
  styles,
  busy,
  indicatorColor,
  ...props
}) => {
  const [animatedValue] = useState(() => new Animated.Value(toValue(busy)));

  useEffect(() => {
    Animated.timing(animatedValue, {
      toValue: toValue(busy),
      duration: 800,
      useNativeDriver: true,
      easing: Easing.out(Easing.exp),
    }).start();
  }, [animatedValue, busy]);

  const buttonAnimation = {
    opacity: animatedValue.interpolate({
      inputRange: [0, 1],
      outputRange: [1, 0],
    }),
    transform: [
      {
        perspective: 1000,
      },
      {
        rotateX: animatedValue.interpolate({
          inputRange: [0, 1],
          outputRange: ['0deg', '-90deg'],
        }),
      },
      {
        translateY: animatedValue.interpolate({
          inputRange: [0, 1],
          outputRange: [0, 35],
        }),
      },
    ],
  };

  const indicatorAnimation = {
    opacity: animatedValue.interpolate({
      inputRange: [0, 1],
      outputRange: [0, 1],
    }),
    transform: [
      {
        perspective: 1000,
      },
      {
        rotateX: animatedValue.interpolate({
          inputRange: [0, 1],
          outputRange: ['90deg', '0deg'],
        }),
      },
      {
        translateY: animatedValue.interpolate({
          inputRange: [0, 1],
          outputRange: [-35, 0],
        }),
      },
    ],
  };

  return (
    <View>
      <Animated.View style={[styles.indicator, indicatorAnimation]}>
        {/* TODO BUG:
          The animating prop has no effect on Android when
          the component initialized with animating=false
          https://github.com/facebook/react-native/issues/9023
        */}
        {busy && <ActivityIndicator size="small" color={indicatorColor} />}
      </Animated.View>
      <Animated.View style={buttonAnimation}>
        <Button themeKey="ActivityButtonButton" {...props} />
      </Animated.View>
    </View>
  );
};

const createStyleSheet = () =>
  StyleSheet.create({
    indicator: {
      ...StyleSheet.absoluteFillObject,
      alignItems: 'center',
      justifyContent: 'center',
    },
  });

export type Props = ActivityButtonProps & WithThemeProps;

export default withTheme(ActivityButton, createStyleSheet, 'ActivityButton');
