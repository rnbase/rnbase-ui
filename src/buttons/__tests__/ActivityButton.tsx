import React from 'react';
import { Animated as AnimatedObj, Text } from 'react-native';

// Note: test renderer must be required after react-native.
import TestRenderer, { act } from 'react-test-renderer';

import { Props } from '../ActivityButton';

jest.doMock('react-native/Libraries/Animated/src/Animated', () => {
  const Animated = jest.requireActual('react-native/Libraries/Animated/src/Animated');

  return {
    ...Animated,
    timing: jest.fn((value: any, { toValue }: any) => {
      value.setValue(toValue);

      return {
        start: jest.fn(),
      };
    }),
  };
});

const Button = (props: any) => React.createElement('Button', props);

jest.doMock('../Button', () => Button);

/**
 * Under test
 */
const { default: ActivityButton } = jest.requireActual('../ActivityButton');

jest.mock('react-native/Libraries/Animated/src/Easing', () => ({
  in: (v: any) => `Easing.in(${v})`,
  out: (v: any) => `Easing.out(${v})`,
  exp: 'Easing.exp',
}));

jest.mock('../Button', () => 'Button');

const Animated: any = AnimatedObj;

const createElement = (props: Props) => (
  <ActivityButton {...props}>
    <Text>TEXT</Text>
  </ActivityButton>
);

const createRenderer = (props: Props) => TestRenderer.create(createElement(props));

it('should render normally', () => {
  expect(createRenderer({})).toMatchSnapshot();
});

it('should render as busy', () => {
  expect(createRenderer({ busy: true })).toMatchSnapshot();
});

it('should render as disabled', () => {
  expect(createRenderer({ disabled: true })).toMatchSnapshot();
});

it('should render with custom styles', () => {
  expect(
    createRenderer({
      style: { backgroundColor: '#000000' },
      textStyle: { color: '#ffffff' },
    }),
  ).toMatchSnapshot();
});

it('should call onPress', () => {
  const onPress = jest.fn();
  const tree = createRenderer({ onPress });

  const button = tree.root.findByType(Button);

  button.props.onPress();

  expect(onPress).toBeCalled();
});

describe('busy animation', () => {
  const start = jest.fn();

  Animated.timing.mockImplementation(() => ({ start }));

  beforeEach(() => {
    start.mockClear();
  });

  it('should not animate', () => {
    const tree = createRenderer({});

    act(() => {});

    Animated.timing.mockClear();

    act(() => {
      tree.update(createElement({ indicatorColor: '#ffffff' }));
    });

    expect(Animated.timing).not.toBeCalled();
  });

  it('should animate to busy', () => {
    const tree = createRenderer({});

    act(() => {});

    Animated.timing.mockClear();

    act(() => {
      tree.update(createElement({ busy: true }));
    });

    expect(Animated.timing).toBeCalledWith(expect.objectContaining({ _value: 0 }), {
      toValue: 1,
      duration: 800,
      useNativeDriver: true,
      easing: 'Easing.out(Easing.exp)',
    });
    expect(start).toBeCalled();
  });

  it('should animate from busy', () => {
    const tree = createRenderer({ busy: true });

    act(() => {});

    Animated.timing.mockClear();

    act(() => {
      tree.update(createElement({ busy: false }));
    });

    expect(Animated.timing).toBeCalledWith(expect.objectContaining({ _value: 1 }), {
      toValue: 0,
      duration: 800,
      useNativeDriver: true,
      easing: 'Easing.out(Easing.exp)',
    });
    expect(start).toBeCalled();
  });
});
