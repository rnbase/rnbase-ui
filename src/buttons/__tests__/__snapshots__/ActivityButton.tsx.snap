// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as busy 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "opacity": 1,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "perspective": 1000,
          },
          Object {
            "rotateX": "0deg",
          },
          Object {
            "translateY": 0,
          },
        ],
      }
    }
  >
    <ActivityIndicator
      animating={true}
      color="#999999"
      hidesWhenStopped={true}
      size="small"
    />
  </View>
  <View
    style={
      Object {
        "opacity": 0,
        "transform": Array [
          Object {
            "perspective": 1000,
          },
          Object {
            "rotateX": "-90deg",
          },
          Object {
            "translateY": 35,
          },
        ],
      }
    }
  >
    <Button>
      <Text>
        TEXT
      </Text>
    </Button>
  </View>
</View>
`;

exports[`should render as disabled 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "perspective": 1000,
          },
          Object {
            "rotateX": "90deg",
          },
          Object {
            "translateY": -35,
          },
        ],
      }
    }
  />
  <View
    style={
      Object {
        "opacity": 1,
        "transform": Array [
          Object {
            "perspective": 1000,
          },
          Object {
            "rotateX": "0deg",
          },
          Object {
            "translateY": 0,
          },
        ],
      }
    }
  >
    <Button
      disabled={true}
    >
      <Text>
        TEXT
      </Text>
    </Button>
  </View>
</View>
`;

exports[`should render normally 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "perspective": 1000,
          },
          Object {
            "rotateX": "90deg",
          },
          Object {
            "translateY": -35,
          },
        ],
      }
    }
  />
  <View
    style={
      Object {
        "opacity": 1,
        "transform": Array [
          Object {
            "perspective": 1000,
          },
          Object {
            "rotateX": "0deg",
          },
          Object {
            "translateY": 0,
          },
        ],
      }
    }
  >
    <Button>
      <Text>
        TEXT
      </Text>
    </Button>
  </View>
</View>
`;

exports[`should render with custom styles 1`] = `
<View>
  <View
    style={
      Object {
        "alignItems": "center",
        "bottom": 0,
        "justifyContent": "center",
        "left": 0,
        "opacity": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "transform": Array [
          Object {
            "perspective": 1000,
          },
          Object {
            "rotateX": "90deg",
          },
          Object {
            "translateY": -35,
          },
        ],
      }
    }
  />
  <View
    style={
      Object {
        "opacity": 1,
        "transform": Array [
          Object {
            "perspective": 1000,
          },
          Object {
            "rotateX": "0deg",
          },
          Object {
            "translateY": 0,
          },
        ],
      }
    }
  >
    <Button
      style={
        Object {
          "backgroundColor": "#000000",
        }
      }
      textStyle={
        Object {
          "color": "#ffffff",
        }
      }
    >
      <Text>
        TEXT
      </Text>
    </Button>
  </View>
</View>
`;
